GC Tombstone: Add allowTombstone request header to enable apps to recover documents from GC-blocking bugs

markfields opened the pull request:
## Description

tl;dr - Adds support for a request header `allowTombstone` that, when passed as true, will enable callers to request a tombstoned object without failing.

_Context: GC Tombstones are objects that GC knows are safe to delete, but rather than deleting them we just mark them as such. This "Tombstone Phase" of rolling out GC allows an app to confirm that there are no GC-blocking bugs in production without exposing their users to permanent data loss._

In order for a host to effectively recover when a GC-impacting bug is found during the Tombstone phase of GC, we're adding a way (via a request header) for app code to request a tombstoned object and receive it back without error.  At this point they can repair whatever incorrect reference was in play that cause the object to be wrongly tombstoned, and in subsequent summaries the object will no longer be a tombstone and the document will be fine.

Fixes [AB#2933](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2933)


## Reviewer Guidance

...

markfields committed:
Add failing tests (and update existing a bit)

markfields committed:
Implement it.  Tests pass

markfields committed:
Refactor logging

markfields committed:
Add handling for blobs too.

And revert/adjust logging changes

agarwal-navin commented on packages/runtime/container-runtime/src/blobManager.ts:
```diff
@@ -268,16 +268,23 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
+                {
+                    eventName: "GC_Tombstone_Blob_Requested",
+                },
+                shouldFail /* logAsError */,
```
We should also log whether `allowTombstone` was set to give us an idea of when this called for recovery purposes.

agarwal-navin commented on packages/runtime/container-runtime/src/blobManager.ts:
```diff
@@ -268,16 +268,23 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
+                error,
+            );
+            if (shouldFail) {
                 throw error;
```
This error will be returned by handle.get(). We need to make sure that consumers can look at it and know that this is caused by tombstone and re-try with the header.

markfields replied:
I'm gonna revert blob stuff for now.  Good catch though

markfields replied:
Tombstone errors thrown to (through) handle.get will include a headers object like `{ isTombstoned: true }`

agarwal-navin commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -430,33 +430,35 @@ export class DataStores implements IDisposable {
+                this.mc.logger,
+                {
+                    eventName: "GC_Tombstone_DataStore_Requested",
+                    viaHandle: headerData.viaHandle,
```
Let's just log `...headerData` so we can also tell if allowTombstone as set or not.

agarwal-navin commented on packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts:
```diff
@@ -3,29 +3,21 @@
+    error: unknown,
 ) {
-    const throwOnTombstoneUsage = event.throwOnTombstoneUsage = mc.config.getBoolean(throwOnTombstoneUsageKey) ?? false;
+    event.category = logAsError ? "error" : "generic";
```
nit: It bugs me that this function accepts event along with properties that are added to event. It would be better to pass in all the properties directly and create an event in this function.

agarwal-navin commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -775,20 +775,19 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
+                error,
+            );
             if (this.throwOnTombstoneUsage) {
                 throw error;
```
We should update this to not throw right? Otherwise, even if data stores are revived by using the recovery scenario, using them will result in this error being thrown. Do you plan to do that in a separate PR?

markfields replied:
Yeah gonna raise/merge that PR first

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -507,13 +531,14 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
-            assert(response.status === 404);
-            assert(response.value === `Datastore removed by gc: ${unreferencedId}`);
+            // Note: if a user makes a request that looks like this, we will also think that the request is via handle
+            await assert.rejects(
```
What about recovery in this case? It could be our bug or app's but in either case, they should be able to recover right?

markfields replied:
Recovery is covered in another test (Can un-tombstone datastores by storing a handle)

markfields committed:
Separate feature flag for request v. use tombstone

markfields committed:
WIP test changes

markfields committed:
merge origin/main

markfields committed:
Actually merge

markfields committed:
Revert blob stuff for now - will follow up with another PR later

markfields committed:
Wrote failing tests for tombstone info set on response/error

markfields committed:
Tests pass

Had to switch the spec on two things:
- successful responses do NOT include the tombstone header
- error thrown by handle.get has tombstone flag in a headers object

markfields committed:
Finish tests and some cleanup

markfields committed:
Clarify expected logs for some tombstone tests

markfields committed:
Some more cleanup

markfields committed:
Fix how header handling is implemented

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -524,6 +524,30 @@ export enum RuntimeHeaders {
+/**
+ * The full set of parsed header data that may be found on Runtime requests
+ */
+export interface RuntimeHeaderData {
+    wait?: boolean;
+    externalRequest?: boolean;
+    viaHandle?: boolean;
+    allowTombstone?: boolean;
+}
```
Is this expanding our public api surface?

markfields replied:
Nope, see index.ts

tyler-cai-microsoft commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -450,23 +462,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
 
-        // If this test starts failing due to runtime is closed errors try first adjusting `sweepTimeoutMs` above
-        itExpects("Requesting tombstoned datastores fails in summarizing container loaded before sweep timeout",
+        // SKIPPED: This test is redundant with the previous one; SweepReady calculation/timers has coverage elsewhere
```
Can we just delete these tests? Also, can we just delete these tests in a separate PR? This doesn't seem like a necessary change we need to backport

tyler-cai-microsoft commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -450,23 +462,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
-        // If this test starts failing due to runtime is closed errors try first adjusting `sweepTimeoutMs` above
-        itExpects("Requesting tombstoned datastores fails in summarizing container loaded before sweep timeout",
+        // SKIPPED: This test is redundant with the previous one; SweepReady calculation/timers has coverage elsewhere
+        itExpects.skip("Requesting tombstoned datastores fails in summarizing container loaded before sweep timeout",
```
Can we just delete these tests?

markfields replied:
It screws up the diff pretty bad and I wanted clear feedback on skipping them so went this way

markfields committed:
Fix tests

markfields committed:
rename props on error to underlyingResponseHeaders

agarwal-navin commented on packages/runtime/runtime-utils/src/dataStoreHelpers.ts:
```diff
@@ -23,6 +23,7 @@ interface IResponseException extends Error {
     message: string;
     code: number;
     stack?: string;
+    underlyingResponseHeaders?: { [key: string]: any; };
```
Should we just call them "headers" similar to how IResponse has headers? It is after all a response exception.

markfields replied:
That's how it was, I changed it last night. Really, handle.get should expose its own shape of error and not expose the fact that internally it's making a request. But I don't want to design that right now. 

I can switch it back to headers if you care. "It's a response exception" - well there's nothing about the actual data that says that, just the TS type. Probably should have errorType but again not touching that here.

agarwal-navin replied:
I don't care that much :-). It's not a big deal as its internal.
Agree with you on handle.get() errors. We should define the error type. Hopefully with the handle work, this will get done. We can create an issue for now.

agarwal-navin commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -430,31 +430,39 @@ export class DataStores implements IDisposable {
-            const error = responseToException(createResponseError(404, "Datastore removed by gc", request), request);
+            const error = responseToException(createResponseError(
+                404,
+                "Datastore removed by gc",
```
We should rename this to "Datastore was deleted". No need to mention "gc" here as it sounds like gc did something wrong :-).

markfields replied:
(Pending)

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -450,23 +468,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
-                async () => requestFluidObject<ITestDataObject>(summarizingContainer, unreferencedId),
-                `Should be able to retrieve a tombstoned datastore in summarizer clients`,
-            );
+            const tombstoneErrorResponse = await containerRuntime_resolveHandle(container, { url: unreferencedId });
```
Can you add comments here explaining why this fails and the next request passes.

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -587,21 +609,17 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
+                url: unreferencedId,
+                headers: { [AllowTombstoneRequestHeaderKey]: true },
+            };
+            const tombstonedObject = await requestFluidObject<ITestDataObject>(tombstoneContainer, requestAllowingTombstone);
```
Why use requestFluidObject here and containerRuntime_resolveHandle above?

markfields replied:
I think anywhere that was getting the handle off the data object I used RequestFluidObject rather than writing that Provider code myself. It's a little arbitrary, definitely open to input.

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -521,17 +536,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
+            }
+            assert.equal(tombstoneError?.code, 404, "Tombstone error from handle.get should have 404 status code");
+            assert.equal(tombstoneError?.message, `Datastore removed by gc: ${unreferencedId}`, "Incorrect message for Tombstone error from handle.get");
+            assert.equal(tombstoneError?.underlyingResponseHeaders?.[TombstoneResponseHeaderKey], true, "Tombstone error from handle.get should include the tombstone flag");
```
Is there any value in adding retrieving this object via resolveHandle after this to mimic what the actual usage would be?

markfields replied:
Ok sure

markfields committed:
PR feedback
